modules/core/src/main/java/org/apache/ignite/internal/worker/CriticalSingleThreadExecutor.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.worker;
+
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Single thread executor instrumented to be used as a {@link CriticalWorker} and being monitorable by the {@link CriticalWorkerWatchdog}.
+ * Must be registered with the watchdog explicitly.
+ */
+public class CriticalSingleThreadExecutor extends ThreadPoolExecutor implements CriticalWorker {
+
+    private Thread lastSeenThread;
```

Comment:
We write to this field before writing to volatile`heartbeatNanos`, and we read from `heartbeatNanos` before reading from `lastSeenThread`, so happens-before is guaranteed. It's a bit fragile because we have to rely on the watchdog doing reads in this order.

I made this field non-volatile to try reducing the number of volatile writes on each job execution (from 2 to 1).

Do you think it's not worth it?

Reply:
I really don't like the fact that we rely on the external party to always call `heartbeatNanos` before `threadId`. I prefer such happens-before relationships to be encapsulated or at least explicitly specified in the javadoc.